src/native/libs/System.Native/pal_interfaceaddresses.c

```diff
@@ -100,12 +101,66 @@ static inline uint8_t mask2prefix(uint8_t* mask, int length)
     return len;
 }
 
+#if !HAVE_IFADDRS && TARGET_ANDROID
+// This structure is exactly the same as struct ifaddrs defined in ifaddrs.h but since the header
+// might not be available (e.g., in bionics used in Android before API 24) we need to mirror it here
+// so that we can dynamically load the getifaddrs function and use it.
+struct ifaddrs
+{
+	struct ifaddrs *ifa_next;
+	char *ifa_name;
+	unsigned int ifa_flags;
+	struct sockaddr *ifa_addr;
+	struct sockaddr *ifa_netmask;
+	union
+	{
+		struct sockaddr *ifu_broadaddr;
+		struct sockaddr *ifu_dstaddr;
+	} ifa_ifu;
+	void *ifa_data;
+};
+#endif
+
+#if !HAVE_GETIFADDRS && TARGET_ANDROID
```

Comment:
this can be further constrained to just < 24:

```suggestion
#if !HAVE_GETIFADDRS && TARGET_ANDROID && __ANDROID_API__ < 24
```

Reply:
and to make it even cleaner we can extract this into a separate define, e.g. `ANDROID_GETIFADDRS_WORKAROUND`, that we can either set here or in the CMakeLists.txt.

Reply:
If I understand it correctly, right now we're always building [System.Native for Android API 21](https://github.com/dotnet/runtime/blob/main/eng/native/build-commons.sh#L88). Once we bump that to 24+, `HAVE_GETIFADDRS` will be 1 and the code won't be included and compiled anyway. We could add `#error Android workaround is not necessary anymore` when `HAVE_GETIFADDRS` is `1` or `__ANDROID_API__ >= 24` but that's maybe too much 😄

Reply:
I extracted the define, but `__ANDROID_API__` isn't defined at the point where it's defined in the CMakeLists.txt and also I believe it's reduntant since checking `HAVE_GETIFADDRS` is equivalent to API >= 24.